src/common/uilocale.cpp

```diff
@@ -103,82 +203,209 @@ wxLocaleIdent wxLocaleIdent::FromTag(const wxString& tag)
         case 2:
         case 3:
             // Either an ISO 3166-1 or UN M.49 region code.
-            locId.m_region = *it;
+            locId.m_region = (*it).Upper();
             break;
 
         case 4:
             // Must be an ISO 15924 script.
-            locId.m_script = *it;
+            wxASSERT_MSG(locId.m_script.empty(), "script already set");
+            locId.m_script = (*it).Left(1).Upper() + (*it).Mid(2).Lower();
             break;
 
         default:
             // This looks to be completely invalid.
+            wxFAIL_MSG("invalid code length, script or region code expected");
             return wxLocaleIdent();
     }
 
-    // If we got the language and the region, we can't parse anything else
-    // (variants, extensions, private use) anyhow.
-    if ( !locId.m_region.empty() )
+    // Check whether we have got the region above.
+    // If not, we must have got the script. So, check if we have the region, too.
+    if (++it == parts.end())
+    {
+        CheckLanguageVariant(locId);
         return locId;
+    }
 
-    // Otherwise we must have got the script above, so check if we have the
-    // region too.
-    if ( ++it == parts.end() )
-        return locId;
+    if (locId.m_region.empty())
+    {
+        switch (it->length())
+        {
+            case 2:
+            case 3:
+                locId.m_region = (*it).Upper(); ++it;
+                break;
+        }
+    }
 
-    switch ( it->length() )
+    // If there is still anything to parse (variants, extensions, private use),
+    // we assign it to the extension.
+    if (it != parts.end())
     {
-        case 2:
-        case 3:
-            locId.m_region = *it;
-            break;
+        wxString custom = *it;
+        while (++it != parts.end())
+        {
+            custom << "-" << *it;
+        }
+        locId.m_extension = custom;
     }
 
+    // We also handle the only language variant known at the time of writing:
+    // valencia (ca-ES-valencia resp ca_ES@valencia).
+    CheckLanguageVariant(locId);
+
     return locId;
 }
 
 wxLocaleIdent& wxLocaleIdent::Language(const wxString& language)
 {
-    m_language = language;
+    if (language.IsSameAs("C", false) || language.IsSameAs("POSIX", false))
+    {
+        m_language = language.Upper();
+    }
+    else if ((language.length() == 2 || language.length() == 3) &&
+             language.find_first_not_of(validCharsAlpha) == wxString::npos)
+    {
+        m_language = language.Lower();
+    }
+    else
+    {
+        m_language = wxEmptyString;
```

Comment:
This is minor and doesn't matter much here, but `wxEmptyString` is an explicit pessimization, as it uses `wxString(const wxChar*)` ctor instead of the default ctor. Using `m_language.clear()` is the most efficient way to clear a string, but if assignment is preferred for aesthetic reasons, I'd rather write
```suggestion
        m_language = wxString();
```

Reply:
`wxEmptyString` is used 5 times in the code. I'll change the code to use method `clear()`.